File: MergeSort.cpp
// Time O(n*logn)
// Space Complexity O(n)

#include <iostream>
#include <vector>
using namespace std;

void Merge(vector<int> &left, vector<int> &right, vector<int> &vec){
    int i{}; int j{}; int k{};
    while(i<left.size() && j<right.size()){
        if(left[i]>right[j]) {
            vec[k] = right[j];
            j++;
        }
        else{
            vec[k] = left[i];
            i++;
        }
        k+=1;
    }
    while(i<left.size()){
        vec[k] = left[i];
        i++; k++;
    }
    while(j<right.size()){
        vec[k] = right[j];
        j++; k++;
    }
}

void Merge_Sort(vector<int> &vec) {
    int n = vec.size();
    if (n < 2)
        return;
    else {
        vector<int> left;
        vector<int> right;
        for (int i = 0; i < n / 2; i++) {
            left.push_back(vec[i]);
        }
        for (int i = n / 2; i < n; i++) {
            right.push_back(vec[i]);
        }

        Merge_Sort(left);
        Merge_Sort(right);
        Merge(left, right, vec);
    }
}

int main(){
    vector<int> vec = {1,4,1,6,8,5,3,7};
    Merge_Sort(vec);
    for(auto i: vec)
      cout<<i<<" ";
}

File: Algorithms/Merge-Sort.py
def merge(L, R, A):
    i = 0
    j = 0
    for k in range(len(A)):
        if i >= len(L):
            A[k] = R[j]
            j += 1
        elif j >= len(R):
            A[k] = L[i]
            i += 1
        elif L[i] <= R[j]:
            A[k] = L[i]
            i += 1
        else:
            A[k] = R[j]
            j += 1
    return A


def merge_sort(A):
    n = len(A)
    if n < 2:
        return A
    else:
        mid = n/2
        L = A[:mid]
        R = A[mid:]
        merge_sort(L)
        merge_sort(R)
        merge(L, R, A)
        return A


print merge_sort([5, 6, 5, 11, 1, 9])
